core/drivers/versal_mbox.c

```diff
@@ -0,0 +1,377 @@
+	[20] = "EFUSE_READ_DNA",
+	[21] = "EFUSE_READ_PUF_USER_FUSES",
+	[22] = "EFUSE_READ_PUF",
+	[23] = "EFUSE_INVALID",
+};
+
+static const char *const crypto_id[] = {
+	[0] = "FEATURES",
+	[1] = "RSA_SIGN_VERIFY",
+	[2] = "RSA_PUBLIC_ENCRYPT",
+	[3] = "RSA_PRIVATE_DECRYPT",
+	[4] = "RSA_KAT",
+	[32] = "SHA3_UPDATE",
+	[33] = "SHA3_KAT",
+	[64] = "ELLIPTIC_GENERATE_PUBLIC_KEY",
+	[65] = "ELLIPTIC_GENERATE_SIGN",
+	[66] = "ELLIPTIC_VALIDATE_PUBLIC_KEY",
+	[67] = "ELLIPTIC_VERIFY_SIGN",
+	[68] = "ELLIPTIC_KAT",
+	[96] = "AES_INIT",
+	[97] = "AES_OP_INIT",
+	[98] = "AES_UPDATE_AAD",
+	[99] = "AES_ENCRYPT_UPDATE",
+	[100] = "AES_ENCRYPT_FINAL",
+	[101] = "AES_DECRYPT_UPDATE",
+	[102] = "AES_DECRYPT_FINAL",
+	[103] = "AES_KEY_ZERO",
+	[104] = "AES_WRITE_KEY",
+	[105] = "AES_LOCK_USER_KEY",
+	[106] = "AES_KEK_DECRYPT",
+	[107] = "AES_SET_DPA_CM",
+	[108] = "AES_DECRYPT_KAT",
+	[109] = "AES_DECRYPT_CM_KAT",
+	[110] = "MAX",
+};
+
+static const char *const puf_id[] = {
+	[0] = "PUF_API_FEATURES",
+	[1] = "PUF_REGISTRATION",
+	[2] = "PUF_REGENERATION",
+	[3] = "PUF_CLEAR_PUF_ID",
+};
+
+static const char *const module[] = {
+	[5] = "CRYPTO",
+	[7] = "FPGA",
+	[11] = "NVM",
+	[12] = "PUF",
+};
+
+static const char *const fpga_id[] = {
+	[1] = "LOAD",
+};
+
+static void versal_mbox_call_trace(uint32_t call)
+{
+	uint32_t mid = call >>  8 & 0xff;
+	uint32_t api = call & 0xff;
+	const char *val = NULL;
+
+	switch (mid) {
+	case 5:
+		if (api < ARRAY_SIZE(crypto_id))
+			val = crypto_id[api];
+
+		break;
+	case 7:
+		if (api < ARRAY_SIZE(fpga_id))
+			val = fpga_id[api];
+
+		break;
+	case 11:
+		if (api < ARRAY_SIZE(nvm_id))
+			val = nvm_id[api];
+
+		break;
+	case 12:
+		if (api < ARRAY_SIZE(puf_id))
+			val = puf_id[api];
+
+		break;
+	default:
+		val = "Invalid";
+	}
+
+	IMSG("--- mbox: service: %s\t call: %s", module[mid],
+	     val ? val : "Invalid");
+};
+
+static TEE_Result mbox_call(enum ipi_api_id id, uint32_t blocking_call)
+{
+	struct thread_smc_args args = {
+		.a0 = PM_SIP_SVC | id,
+		.a1 = reg_pair_to_64(0, ipi.lcl),
+		.a2 = reg_pair_to_64(0, ipi.rmt),
+		.a3 = reg_pair_to_64(0, blocking_call),
+	};
+
+	thread_smccc(&args);
+
+	/* give the PLM time to access the console */
+	if (IS_ENABLED(CFG_VERSAL_TRACE_PLM))
+		mdelay(500);
+
+	if (args.a0)
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result versal_mbox_write_req(struct ipi_cmd *cmd)
+{
+	size_t i = 0;
+
+	for (i = 0; i < MAX_IPI_BUF; i++) {
+		if (!cmd->ibuf[i].mem.buf)
+			continue;
+
+		if (!IS_ALIGNED((uintptr_t)cmd->ibuf[i].mem.buf,
+				CACHELINE_LEN)) {
+			EMSG("address not aligned: buffer %ld - 0x%p", i,
+			     cmd->ibuf[i].mem.buf);
+			return TEE_ERROR_GENERIC;
+		}
+
+		if (!IS_ALIGNED(cmd->ibuf[i].mem.alloc_len, CACHELINE_LEN)) {
+			EMSG("length not aligned: buffer %ld - %ld",
+			     i, cmd->ibuf[i].mem.alloc_len);
+			return TEE_ERROR_GENERIC;
+		}
+
+		cache_operation(TEE_CACHEFLUSH,
+				cmd->ibuf[i].mem.buf,
```

Comment:
Fits on the previous line I think

Reply:
I would have preferred the way it was (personal preference for readability) but ok